ingester/series_map.go

```diff
@@ -0,0 +1,88 @@
+package ingester
+
+import (
+	"sync"
+
+	"github.com/prometheus/common/model"
+)
+
+// fingerprintSeriesPair pairs a fingerprint with a memorySeries pointer.
+type fingerprintSeriesPair struct {
+	fp     model.Fingerprint
+	series *memorySeries
+}
+
+// seriesMap maps fingerprints to memory series. All its methods are
+// goroutine-safe. A seriesMap is effectively is a goroutine-safe version of
+// map[model.Fingerprint]*memorySeries.
+type seriesMap struct {
+	mtx sync.RWMutex
+	m   map[model.Fingerprint]*memorySeries
+}
+
+// newSeriesMap returns a newly allocated empty seriesMap. To create a seriesMap
+// based on a prefilled map, use an explicit initializer.
+func newSeriesMap() *seriesMap {
+	return &seriesMap{m: make(map[model.Fingerprint]*memorySeries)}
+}
+
+// length returns the number of mappings in the seriesMap.
+func (sm *seriesMap) length() int {
+	sm.mtx.RLock()
+	defer sm.mtx.RUnlock()
+
+	return len(sm.m)
+}
+
+// get returns a memorySeries for a fingerprint. Return values have the same
+// semantics as the native Go map.
+func (sm *seriesMap) get(fp model.Fingerprint) (s *memorySeries, ok bool) {
+	sm.mtx.RLock()
+	s, ok = sm.m[fp]
+	// Note that the RUnlock is not done via defer for performance reasons.
+	// TODO(beorn7): Once https://github.com/golang/go/issues/14939 is
+	// fixed, revert to the usual defer idiom.
+	sm.mtx.RUnlock()
+	return
+}
+
+// put adds a mapping to the seriesMap. It panics if s == nil.
+func (sm *seriesMap) put(fp model.Fingerprint, s *memorySeries) {
+	sm.mtx.Lock()
+	defer sm.mtx.Unlock()
+
+	if s == nil {
+		panic("tried to add nil pointer to seriesMap")
+	}
+	sm.m[fp] = s
+}
+
+// del removes a mapping from the series Map.
+func (sm *seriesMap) del(fp model.Fingerprint) {
+	sm.mtx.Lock()
+	defer sm.mtx.Unlock()
+
+	delete(sm.m, fp)
+}
+
+// iter returns a channel that produces all mappings in the seriesMap. The
+// channel will be closed once all fingerprints have been received. Not
+// consuming all fingerprints from the channel will leak a goroutine. The
+// semantics of concurrent modification of seriesMap is the similar as the one
+// for iterating over a map with a 'range' clause. However, if the next element
+// in iteration order is removed after the current element has been received
+// from the channel, it will still be produced by the channel.
+func (sm *seriesMap) iter() <-chan fingerprintSeriesPair {
```

Comment:
I guess this is just being moved, but I don't really understand why we need it. If it's there to let us get the contents in a safe way, a copy would be more clear, probably not less performant. If it's there to allow us to fiddle with the data in a `seriesMap`, it should just take a callback. Either way no goroutine needed, which means less chance of goroutine leaks from irresponsible callers.